File: 206_reverse_linked_list/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        return self.reverse(None, head)

    def reverse(self, pre: ListNode, cur: ListNode):
        if not cur: return pre

        next = cur.next
        cur.next = pre

        return self.reverse(cur, next)

    def reverseList2(self, head: ListNode) -> ListNode:
        cur, pre = head, None
        while cur:
            cur.next, pre, cur = pre, cur, cur.next

        return pre


if __name__ == '__main__':
    head = ListNode(0)
    cur = head
    for i in range(1, 10):
        cur.next = ListNode(i)
        cur = cur.next

    solution = Solution()
    cur = solution.reverseList2(head)

    while cur:
        print(cur.val, end = " ")
        cur = cur.next

File: 206_reverse_linked_list/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
    }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        return reverse(null, head);
    }

    public ListNode reverse(ListNode pre, ListNode cur) {
        if (cur == null) return pre;
        ListNode next = cur.next;
        cur.next = pre;

        return reverse(cur, next);
    }

    public ListNode reverseList2(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode pre = null;
        ListNode cur = head;
        ListNode next = null;

        while (cur != null) {
            next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }

        return pre;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(0);
        ListNode cur = head;

        for (int i = 1; i < 10; i++) {
            cur.next = new ListNode(i);
            cur = cur.next;
        }

        Solution solution = new Solution();
        cur = solution.reverseList2(head);

        while (cur != null) {
            System.out.print(cur.val + " ");
            cur = cur.next;
        }
    }
}
